File: src/main/java/com/github/zoewithabang/command/ListBotMessages.java
package com.github.zoewithabang.command;

import com.github.zoewithabang.bot.IBot;
import com.github.zoewithabang.model.BotMessage;
import com.github.zoewithabang.service.BotMessageService;
import com.github.zoewithabang.service.OptionService;
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageReceivedEvent;
import sx.blah.discord.handle.obj.IChannel;
import sx.blah.discord.util.EmbedBuilder;

import java.sql.SQLException;
import java.util.List;
import java.util.Properties;

public class ListBotMessages implements ICommand
{
    public static final String COMMAND = "messages";
    private IBot bot;
    private Properties botProperties;
    private String prefix;
    private OptionService optionService;
    private BotMessageService botMessageService;
    
    public ListBotMessages(IBot bot, Properties botProperties)
    {
        this.bot = bot;
        this.botProperties = botProperties;
        prefix = botProperties.getProperty("prefix");
        optionService = new OptionService(botProperties);
        botMessageService = new BotMessageService(botProperties);
    }
    
    @Override
    public void execute(MessageReceivedEvent event, List<String> args, boolean sendBotMessages)
    {
        IChannel eventChannel = event.getChannel();
        List<BotMessage> botMessages;
    
        if(!validateArgs(event, args))
        {
            LOGGER.warn("Validation failed for ListBotMessages.");
            if(sendBotMessages)
            {
                postUsageMessage(eventChannel);
            }
            return;
        }
    
        try
        {
            botMessages = botMessageService.getAllBotMessages();
        }
        catch(SQLException e)
        {
            bot.postErrorMessage(eventChannel, sendBotMessages, COMMAND, 15001);
            return;
        }
    
        postBotMessagesMessage(eventChannel, botMessages);
    }
    
    @Override
    public boolean validateArgs(MessageReceivedEvent event, List<String> args)
    {
        return args.size() == 0;
    }
    
    public void postBotMessagesMessage(IChannel channel, List<BotMessage> botMessages)
    {
        EmbedBuilder builder = new EmbedBuilder();
    
        builder.withAuthorName("List of aliases:");
        builder.withColor(optionService.getBotColour());
    
        for(BotMessage botMessage : botMessages)
        {
            String title = prefix + botMessage.getName();
            String content = "Description: " + botMessage.getDescription();
            builder.appendField(title, content, false);
        }
    
        bot.sendEmbedMessage(channel, builder.build());
    }
    
    @Override
    public void postUsageMessage(IChannel channel)
    {
        String title = prefix + COMMAND;
        String content = "List the currently stored bot messages.";
    
        EmbedBuilder builder = new EmbedBuilder();
        builder.appendField(title, content, false);
        builder.withColor(optionService.getBotColour());
    
        bot.sendEmbedMessage(channel, builder.build());
    }
}

File: src/main/kotlin/com/github/zoewithabang/command/ListBotMessages.kt
package com.github.zoewithabang.command

import com.github.zoewithabang.bot.Bot
import com.github.zoewithabang.model.BotMessage
import com.github.zoewithabang.service.BotMessageService
import com.github.zoewithabang.service.OptionService
import org.apache.logging.log4j.LogManager
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageReceivedEvent
import sx.blah.discord.handle.obj.IChannel
import sx.blah.discord.util.EmbedBuilder

import java.sql.SQLException
import java.util.Properties

class ListBotMessages(private val bot: Bot, private val botProperties: Properties) : Command {
    private val logger = LogManager.getLogger("ListBotMessages")
    private val prefix: String = botProperties.getProperty("prefix")
    private val optionService: OptionService = OptionService(botProperties)
    private val botMessageService: BotMessageService = BotMessageService(botProperties)

    override fun execute(event: MessageReceivedEvent, args: List<String>, sendBotMessages: Boolean) {
        val eventChannel = event.getChannel()
        val botMessages: List<BotMessage>

        if (!validateArgs(event, args)) {
            logger.warn("Validation failed for ListBotMessages.")

            if (sendBotMessages) {
                postUsageMessage(eventChannel)
            }

            return
        }

        try {
            botMessages = botMessageService.allBotMessages
        } catch (e: SQLException) {
            bot.postErrorMessage(eventChannel, sendBotMessages, COMMAND, 15001)

            return
        }

        postBotMessagesMessage(eventChannel, botMessages)
    }

    override fun validateArgs(event: MessageReceivedEvent, args: List<String>): Boolean {
        return args.isEmpty()
    }

    private fun postBotMessagesMessage(channel: IChannel, botMessages: List<BotMessage>) {
        val builder = EmbedBuilder()

        builder.withAuthorName("List of messages:")
        builder.withColor(optionService.botColour)

        for (botMessage in botMessages) {
            val title = prefix + BotSay.COMMAND + " " + botMessage.name
            val content = botMessage.description
            builder.appendField(title, content, false)
        }

        bot.sendEmbedMessage(channel, builder.build())
    }

    override fun postUsageMessage(channel: IChannel) {
        val title = prefix + COMMAND
        val content = "List the currently stored bot messages."

        val builder = EmbedBuilder()
        builder.appendField(title, content, false)
        builder.withColor(optionService.botColour)

        bot.sendEmbedMessage(channel, builder.build())
    }

    companion object {
        const val COMMAND = "messages"
    }
}
